### ISSUE TITLE
Audio filters with same name cannot be removed.

### ISSUE THREAD

**Issue by timlenertz** [iina/iina#3620]
**System and IINA version:**

- macOS 12.1
- IINA 1.2.0 Build 129 (mpv 0.32.0)

**Expected behavior:**
Added two audio filters (both name 'scaletempo', different values). Then they can no longer be removed. When clicking the minus button, the overlay "Filter removed" appears, but the filter is not removed. It stays when reopening IINA.
The only way to remove it is to remove/modify the files in ~/Library/Application Support/com.colliderli.iina/watch_later.

**Actual behavior:**
Clicking the minus button in the audio filters list window should delete the selected audio filter.

<br>

**Steps to reproduce:**
- Open a video file in IINA.
- Add two audio filters name 'scaletempo', one with 'stride=8:overlap=1:search=10:scale=2:speed=pitch', the other with 'stride=8:overlap=1:search=10:scale=0.4:speed=pitch' (for example).
- Try to remove them again.

**How often does this happen?**
First try: both could not be removed. (maybe added more filters before)
Second try: one of the filters could not be removed.

**Comment by low-batt** [comment]
### Analysis

Reproduced for me. This is a defect. This does not require adding multiple filters. IINA can have trouble removing even one filter.

Audio and video filters are retrieved from mpv in `MPVController.getFilters`. That method uses `MPVNode.parse` to convert the `mpv_node` returned by `mpv_get_property` into Swift objects. When converting the objects within the `mpv_node` the method parses a `MPV_FORMAT_NODE_MAP` object into a [Swift Dictionary](https://developer.apple.com/documentation/swift/dictionary). The order of key-value pairs in a Swift Dictionary is unpredictable, so this does not preserve the order of keys in the `MPV_FORMAT_NODE_MAP`. This should not be a problem since the documentation for `MPV_FORMAT_NODE_MAP` explicitly documents that _keys are in random order_.

Trouble is the mpv `af` and `vf` `remove` command when matching filters compares _argument order_. The order of arguments returned by `mpv_get_property` in the `MPV_FORMAT_NODE_MAP` does not appear to be random as indicated in the mpv documentation. In a test the keys matched the order of the arguments when the filter was added. But parsing the map into a Swift Dictionary is changing the order of the filter arguments. This causes the mpv `af` `remove` command to fail to find the filter.

### Fixing

Before proceeding with a fix we need clarification from mpv as to whether the keys in a `MPV_FORMAT_NODE_MAP` can be counted on to be in the proper order. Is the documentation that indicates keys are in random order incorrect?

If the keys are indeed ordered then IINA must be changed to parse a `MPV_FORMAT_NODE_MAP` into a [Swift OrderedDictionary](https://github.com/apple/swift-collections/blob/main/Documentation/OrderedDictionary.md). The filter window worked properly and was able to remove filters in a quick test of such a change. However this change is inappropriate if the keys in a `MPV_FORMAT_NODE_MAP` are unordered as the mpv documentation indicates.

I will be filling an issue with mpv to get a clarification of  `MPV_FORMAT_NODE_MAP` behavior.

**Comment by low-batt** [comment]
### Root Cause

Briefly, the root cause is that certain types of filters accept multiple parameters. Such filters do not require parameters to appear in a particular order. This means there are multiple valid string representations for the same filter. IINA is relying upon being able to find matching filters based on their string representations which can erroneously fail to find a match due to differing string representations.

### Additional Problems

Investigating the problem reported revealed additional filter related problems

To begin with this same issue applies to video filters. The following custom mpv video filter will demonstrate the same issues experienced with audio filters:
```text
Filter name: format
Filter value: convert=no:colorlevels=full:colormatrix=auto:primaries=auto
```

Such filters also trigger problems with IINA's ability to save filters. Add the following custom mpv audio filter:

```text
Filter name: scaletempo
Filter value: overlap=1:search=10:speed=pitch:scale=2:stride=8
```

Then click the `save` button. The filter should appear in the `Saved filters` section with the check box checked. But frequently the box will not be checked. Clicking the box will add the filter again and may or may not result in the box being checked. The more parameters in the filter the more likely this failure will be encountered.

This debug logging shows another problem:
```text
21:54:19.532 [iina][d] rhs Optional(type=nil, name=@iina_flip:vflip, label=nil, params=nil, rawParamString=nil)
```

The `MPVFilter` initializer that parses a string is failing to separate the label from the filter name. 

### Fixing

Experiments appeared to indicate that a mpv `MPV_FORMAT_NODE_MAP` might be a form of linked hash map that preserves insertion order and yet the mpv documentation indicated keys are in random order. In [mpv issue #9841](https://github.com/mpv-player/mpv/issues/9841) the mpv project confirmed the documentation for `MPV_FORMAT_NODE_MAP` is correct and keys **must be** treated as being in **random** order. The preferred way to remove filters is to use labels and refer to a filter using its label, or remove the filter based on its position in the filter list. See that mpv issue for details. Thank you mpv project for the help!

IINA is already using labels for filters it creates. IINA is not able to enforce use of labels for user created filters as users can add filters directly to mpv using the IPC interface as seen here:
```bash
low-batt@gag ~$ echo '{ "command": ["af", "add", "scaletempo:stride=8:overlap=1:search=10:scale=2:speed=pitch"] }' | socat - /tmp/mpv-socket | jq
{
  "data": null,
  "request_id": 0,
  "error": "success"
}
```

Therefore the proposed fix removes filters based on their position within the list of filters.

The commit in the pull request will:
- Add a new method `removeFilter` to `MPVController` that removes a filter based on its position within the list of filters
- Add methods for removing audio and video filters based on their position within the list of filters to `PlayerCore`
- Change `FilterWindowController.removeFilterAction` to use the new methods

The `MPVController.removeFilter` method uses `mpv_get_property` to get the filter list. Then it modifies the list removing the filter and applies the changes using `mpv_set_property`.

This solves the problem reported, but not the issue with saved filters. The commit in the pull request will also:
- Add a `parseRawParamString` method to `MPVFilter`
- Add an `isEqual` method to `MPVFilter`
- Change  `MPVFilter.init(rawString:)` to properly handle filters with labels
- Add a `debugDescription`  property to  `MPVFilter`
- Change `FilterWindowController.toggleSavedFilterAction` to locate the saved filter using `MPVFilter` objects instead of strings
- Change `FilterWindowController.reloadTable` to locate saved filters using `MPVFilter` objects instead of strings
- Add a `filtersAsString` method to `FilterWindowController`
- Add a `debugDescription` property to `SavedFilter`

With these IINA parses the saved filter string and extracts parameters into a `Dictionary`. Then IINA can compare the parameters of two filters by comparing dictionaries instead of comparing string representations.

This is a significant change. An intense mean detailed critical review of PR #3665 is called for.

During the investigation of this issue a custom `debugDescription` property for `mpv_node` that can be used to log an entire node tree for debugging. That code is in the separate PR #3657.

**Comment by low-batt** [comment]
The fix in PR #3665 has been merged into the develop branch.

**Comment by low-batt** [comment]
Closing. Fixed in IINA 1.3.0.
